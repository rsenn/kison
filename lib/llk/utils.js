// @ts-check

const data = require('../data');

let { cachedStateMatchMap } = data;

function setStateCacheMatch(state, tokensLength) {
  return
  let holder = cachedStateMatchMap.get(state);
  if (!holder) {
    holder = new Set();
    cachedStateMatchMap.set(state, holder);
  }
  holder.add(tokensLength);
}

function getStateCacheMatch(state, tokensLength) {
  return;
  // consumed token length may be different!
  const holder = cachedStateMatchMap.get(state);
  return holder?.has(tokensLength);
}

function clearStateMatchCache() {
  return
  cachedStateMatchMap = new Map();
}

module.exports = {
  setStateCacheMatch,
  getStateCacheMatch,
  clearStateMatchCache,
};