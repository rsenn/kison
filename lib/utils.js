/**
 * @ignore
 * utils for kison.
 * @author yiminghe@gmail.com
 */

var util = require("modulex-util");

var doubleReg = /"/g;
var single = /'/g;

const gened = {
  mix(to, from) {
    for (var f in from) {
      to[f] = from[f];
    }
  },

  isArray(obj) {
    return "[object Array]" === Object.prototype.toString.call(obj);
  },

  inArray(item, arr) {
    for (var i = 0, l = arr.length; i < l; i++) {
      if (arr[i] === item) {
        return true;
      }
    }
    return false;
  },

  each(object, fn, context) {
    if (object) {
      var key,
        val,
        length,
        i = 0;

      context = context || null;

      if (!isArray(object)) {
        for (key in object) {
          // can not use hasOwnProperty
          if (fn.call(context, object[key], key, object) === false) {
            break;
          }
        }
      } else {
        length = object.length;
        for (val = object[0]; i < length; val = object[++i]) {
          if (fn.call(context, val, i, object) === false) {
            break;
          }
        }
      }
    }
  }
};

var { isArray } = gened;

const utils = (module.exports = Object.assign({}, util, gened, {
  eachRhs(rhs, fn) {
    utils.each(rhs, (r, index) => {
      if (typeof r !== "string") {
        return;
      }
      fn(r, index);
    });
  },

  assertSymbolString(symbol) {
    if (typeof symbol !== "string") {
      throw new Error("invalid symbol".symbol);
    }
  },

  filterRhs(rhs) {
    return rhs.filter(r => typeof r === "string");
  },

  toFunctionString(fn, name) {
    const str = fn.toString();
    if (str.match(/^function\s+\(/)) {
      return str.replace(/^function\s+\(/, `function ${name || fn.name}(`);
    }
    return "function " + str;
  },

  escapeString(str, quote) {
    var regexp = single;
    if (quote === '"') {
      regexp = doubleReg;
    } else {
      quote = "'";
    }
    return str
      .replace(/\\/g, "\\\\")
      .replace(/\r/g, "\\r")
      .replace(/\n/g, "\\n")
      .replace(/\t/g, "\\t")
      .replace(regexp, "\\" + quote);
  },

  serializeObject(obj, excludeReg) {
    var r;

    if (
      excludeReg &&
      typeof excludeReg === "function" &&
      (r = excludeReg(obj)) === false
    ) {
      return false;
    }

    if (r !== undefined) {
      obj = r;
    }

    var ret = [];

    if (typeof obj === "string") {
      return "'" + utils.escapeString(obj) + "'";
    } else if (typeof obj === "number") {
      return obj + "";
    } else if (util.isRegExp(obj)) {
      return (
        "/" +
        obj.source +
        "/" +
        (obj.global ? "g" : "") +
        (obj.ignoreCase ? "i" : "") +
        (obj.multiline ? "m" : "")
      );
    } else if (util.isArray(obj)) {
      ret.push("[");
      var sub = [];
      util.each(obj, function(v) {
        var t = utils.serializeObject(v, excludeReg);
        if (t !== false) {
          sub.push(t);
        }
      });
      ret.push(sub.join(", "));
      ret.push("]");
      return ret.join("");
    } else if (typeof obj === "object") {
      ret = [];
      ret[0] = "{";
      var start = 1;
      for (var i in obj) {
        var v = obj[i];
        if (excludeReg && util.isRegExp(excludeReg) && i.match(excludeReg)) {
          continue;
        }
        var t = utils.serializeObject(v, excludeReg);
        if (t === false) {
          continue;
        }
        /*jshint quotmark:false*/
        var key = "'" + utils.escapeString(i) + "'";
        ret.push((start ? "" : ",") + key + ": " + t);
        start = 0;
      }
      ret.push("}");
      return ret.join("\n");
    } else if (typeof obj === "function") {
      return (obj + "").replace(/^[\w-]+/, "function");
    } else {
      return obj + "";
    }
  }
}));
