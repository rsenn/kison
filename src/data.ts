import { ProductionRule } from './Grammar';
import type Lexer from './Lexer';
import type { SymbolStateUnit } from './llk/sm';
import { Rhs } from './types';

const gened: {
  VIRTUAL_OPTIONAL_RULE_INDEX: number;
  START_TAG: string;
  smUnitBySymbol: Record<string, SymbolStateUnit>;
  productionSkipAstNodeSet: Set<number> | undefined;
  symbolStack: any[];
  productionsBySymbol: Record<
    string,
    {
      productions: ProductionRule[];
      ruleIndexes: number[];
    }
  >;
  productionAddAstNodeFlag: 1;
  productionEndFlag: 2;
} = {
  VIRTUAL_OPTIONAL_RULE_INDEX: -100,
  START_TAG: '$START',
  smUnitBySymbol: {},
  productionSkipAstNodeSet: undefined,
  symbolStack: [{}],
  productionsBySymbol: {},
  productionAddAstNodeFlag: 1,
  productionEndFlag: 2,
};

export default {
  gened,
  Lexer: {} as typeof Lexer,
  lexer: {} as Lexer,
  productionRuleIndexMap: {} as Record<number, number>,
  parser: {} as {
    userData: any;
    table: any;
    prioritySymbolMap: Record<string, string>;
    getProductionRhs: (p: ProductionRule) => Rhs;
    getProductionSymbol: (p: ProductionRule) => string;
    getProductionLabel: (p: ProductionRule) => string;
    getProductionAction: (p: ProductionRule) => Function | undefined;
    getProductionPredict: (p: ProductionRule) => Function | undefined;
    productions: ProductionRule[];
  },
  ...gened,
};
