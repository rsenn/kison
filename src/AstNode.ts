import type { ParseError } from './types';
import data from './data';

interface Params {
  parent?: AstNode;
}

interface TokenParams extends Params {
  token?: string;
  text?: string;
  t?: string;
}

interface ErrorTokenParams extends TokenParams {
  error?: ParseError;
}

interface SymbolParams extends Params {
  id: number;
  children?: AstNode[];
  symbol?: string;
  label?: string;
  internalRuleIndex?: number;
}

export type AstNode = AstSymbolNode | AstTokenNode;

export class BaseAstNode {
  parent?: AstSymbolNode;
  start: number = 0;
  end: number = 0;
  firstLine: number = 0;
  lastLine: number = 0;
  firstColumn: number = 0;
  lastColumn: number = 0;

  toJSON() {
    const ret: Record<string, any> = {};
    for (const k of Object.keys(this)) {
      if (k !== 'parent' && k !== 't') {
        const v = (this as any)[k];
        if (v !== undefined) {
          ret[k] = v;
        }
      }
    }
    return ret;
  }
}

export class AstTokenNode extends BaseAstNode {
  token: string = '';
  t: string = '';
  recovery?: string;
  type: 'token' = 'token';
  constructor(params: TokenParams) {
    super();
    Object.assign(this, params);
  }
}

export class AstErrorNode extends AstTokenNode {
  error?: ParseError;
  constructor(ErrorTokenParams: ErrorTokenParams) {
    super(ErrorTokenParams);
    Object.assign(ErrorTokenParams);
  }
}

const { productionRuleIndexMap } = data;

export class AstSymbolNode extends BaseAstNode {
  symbol: string = '';
  label?: string;
  type: 'symbol' = 'symbol';
  children: AstNode[] = [];
  ruleIndex: number = -1;
  internalRuleIndex: number = -1;
  constructor(params: SymbolParams) {
    super();
    Object.assign(this, params);
    if (params.children) {
      this.setChildren(params.children);
    }
    if (params.internalRuleIndex !== undefined) {
      this.ruleIndex = productionRuleIndexMap[this.internalRuleIndex];
    }
  }

  addChild(c: AstNode) {
    this.addChildren([c]);
  }

  addChildren(cs: AstNode[]) {
    this.children = this.children || [];
    this.children.push(...cs);
    this.setChildren(this.children);
  }

  refreshChildren() {
    this.setChildren(this.children);
  }

  setChildren(cs: AstNode[]) {
    if (!cs.length) {
      this.children = [];
      return;
    }
    const first = cs[0];
    const last = cs[cs.length - 1];
    this.start = first.start;
    this.end = last.end;
    this.firstLine = first.firstLine;
    this.lastLine = last.lastLine;
    this.firstColumn = first.firstColumn;
    this.lastColumn = last.lastColumn;
    this.children = cs;
    for (const c of cs) {
      c.parent = this;
    }
  }
}
