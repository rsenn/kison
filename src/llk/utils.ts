// @ts-check

import data from '../data';

let { productionsBySymbol } = data;

function isSymbol(s: string) {
  return !!productionsBySymbol[s];
}

export { isSymbol };
