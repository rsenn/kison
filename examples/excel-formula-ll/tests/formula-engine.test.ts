import { utils, FormulaEngine, CellValue } from '../src/index';
import { makeNumber, makeFormula } from './utils';

const { parseCoord, toCoordString } = utils;

describe('excel-formula-engine', () => {
  let g: FormulaEngine = undefined!;

  beforeEach(() => {
    g = new FormulaEngine();

    // h-> g->f->d -> c -> b
    // f->h
    // k->f
    // e->c
    g.initWithValues([
      [
        makeNumber(1), // a1
        makeFormula('a1+1'), // b1
        makeFormula('b1+2'), // c1
        makeFormula('c1+3'), // d1
        makeFormula('c1+4'), // e1
        makeFormula('d1+h1'), // f1
        makeFormula('f1'), // g1
        makeFormula('g1'), // h1
        makeFormula('f1'), // i1
      ],
      [
        makeFormula('sum(a1:c1)'), //a2
        makeNumber(9), // b2
        makeNumber(11), // b2
        makeNumber(13), // b2
      ],
    ]);
  });

  type Cell = { adr: string; value: CellValue };

  function getCellValues() {
    const ret: Cell[][] = [[], []];
    for (let row = 1; row <= g.height; row++) {
      for (let col = 1; col <= g.width; col++) {
        const addr = { row, col };
        const value = g.getCellValue(addr);
        if (value) {
          ret[0].push({
            adr: toCoordString(addr),
            value,
          });
        }
      }
    }
    return ret;
  }

  it('compute works', () => {
    expect(g.width).toBe(9);
    expect(g.height).toBe(2);

    const ret = getCellValues();

    expect(ret).toMatchSnapshot();
  });

  it('insertRow works', () => {
    g.insertRows(1, 2);
    const ret = getCellValues();

    expect(ret).toMatchSnapshot();
  });

  it('recompute works', () => {
    g.setCellValue(parseCoord('c1'), makeNumber(10));

    const ret = getCellValues();

    expect(ret).toMatchSnapshot();
  });

  it('array works', () => {
    function e(formula: string) {
      return {
        formula,
        value: g.evaluateFormula(formula),
      };
    }

    expect(e('A1:A2+B1:B2')).toMatchSnapshot();

    expect(e('A1:A2+B1:C1')).toMatchSnapshot();

    expect(e('A1:B2+C1:D2')).toMatchSnapshot();

    expect(e('9+C1:D2')).toMatchSnapshot();
  });
});
