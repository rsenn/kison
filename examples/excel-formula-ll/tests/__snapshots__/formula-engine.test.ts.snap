// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`excel-formula-engine array works 1`] = `
Object {
  "formula": "A1:A2+B1:B2",
  "value": Object {
    "type": "array",
    "value": Array [
      Array [
        Object {
          "type": "number",
          "value": 3,
        },
      ],
      Array [
        Object {
          "message": "Value Not Available Error!",
          "type": "error",
          "value": "#N/A",
        },
      ],
    ],
  },
}
`;

exports[`excel-formula-engine array works 2`] = `
Object {
  "formula": "A1:A2+B1:C1",
  "value": Object {
    "type": "array",
    "value": Array [
      Array [
        Object {
          "type": "number",
          "value": 3,
        },
        Object {
          "type": "number",
          "value": 5,
        },
      ],
      Array [
        Object {
          "message": "Value Not Available Error!",
          "type": "error",
          "value": "#N/A",
        },
        Object {
          "message": "Value Not Available Error!",
          "type": "error",
          "value": "#N/A",
        },
      ],
    ],
  },
}
`;

exports[`excel-formula-engine array works 3`] = `
Object {
  "formula": "A1:B2+C1:D2",
  "value": Object {
    "type": "array",
    "value": Array [
      Array [
        Object {
          "type": "number",
          "value": 5,
        },
        Object {
          "type": "number",
          "value": 9,
        },
      ],
      Array [
        Object {
          "message": "Value Not Available Error!",
          "type": "error",
          "value": "#N/A",
        },
        Object {
          "message": "Value Not Available Error!",
          "type": "error",
          "value": "#N/A",
        },
      ],
    ],
  },
}
`;

exports[`excel-formula-engine array works 4`] = `
Object {
  "formula": "9+C1:D2",
  "value": Object {
    "type": "array",
    "value": Array [
      Array [
        Object {
          "type": "number",
          "value": 13,
        },
        Object {
          "type": "number",
          "value": 16,
        },
      ],
      Array [
        Object {
          "type": "number",
          "value": 9,
        },
        Object {
          "type": "number",
          "value": 9,
        },
      ],
    ],
  },
}
`;

exports[`excel-formula-engine compute works 3`] = `
Array [
  Array [
    Object {
      "adr": "A1",
      "value": Object {
        "type": "number",
        "value": 1,
      },
    },
    Object {
      "adr": "B1",
      "value": Object {
        "formula": "a1+1",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 2,
        },
      },
    },
    Object {
      "adr": "C1",
      "value": Object {
        "formula": "b1+2",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 4,
        },
      },
    },
    Object {
      "adr": "D1",
      "value": Object {
        "formula": "c1+3",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 7,
        },
      },
    },
    Object {
      "adr": "E1",
      "value": Object {
        "formula": "c1+4",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 8,
        },
      },
    },
    Object {
      "adr": "F1",
      "value": Object {
        "formula": "d1+h1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "G1",
      "value": Object {
        "formula": "f1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "H1",
      "value": Object {
        "formula": "g1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "I1",
      "value": Object {
        "formula": "f1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "A3",
      "value": Object {
        "formula": "sum(a1:c2)",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 7,
        },
      },
    },
    Object {
      "adr": "B3",
      "value": Object {
        "type": "number",
        "value": 9,
      },
    },
    Object {
      "adr": "C3",
      "value": Object {
        "formula": "sum(a1:b3)",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 19,
        },
      },
    },
    Object {
      "adr": "D3",
      "value": Object {
        "type": "number",
        "value": 13,
      },
    },
    Object {
      "adr": "A4",
      "value": Object {
        "formula": "a3+1",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 8,
        },
      },
    },
    Object {
      "adr": "B4",
      "value": Object {
        "formula": "d1+10",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 17,
        },
      },
    },
    Object {
      "adr": "C4",
      "value": Object {
        "formula": "b3+6",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 15,
        },
      },
    },
  ],
  Array [],
]
`;

exports[`excel-formula-engine deleteRows works 1`] = `
Array [
  Object {
    "address": "B2",
    "after": Object {
      "formula": "#REF!+10",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#REF!",
      },
    },
    "before": Object {
      "formula": "d1+10",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 17,
      },
    },
    "type": "update",
  },
  Object {
    "address": "A1",
    "after": Object {
      "formula": "sum(#REF!)",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#REF!",
      },
    },
    "before": Object {
      "formula": "sum(a1:c2)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 7,
      },
    },
    "type": "update",
  },
  Object {
    "address": "A2",
    "after": Object {
      "formula": "A1+1",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#REF!",
      },
    },
    "before": Object {
      "formula": "a3+1",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 8,
      },
    },
    "type": "update",
  },
  Object {
    "address": "C1",
    "after": Object {
      "formula": "sum(A1:B1)",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#REF!",
      },
    },
    "before": Object {
      "formula": "sum(a1:b3)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 19,
      },
    },
    "type": "update",
  },
  Object {
    "address": "C2",
    "after": Object {
      "formula": "B1+6",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 15,
      },
    },
    "before": Object {
      "formula": "b3+6",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 15,
      },
    },
    "type": "update",
  },
]
`;

exports[`excel-formula-engine deleteRows works 2`] = `
Array [
  Array [
    Object {
      "adr": "A1",
      "value": Object {
        "formula": "sum(#REF!)",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#REF!",
        },
      },
    },
    Object {
      "adr": "B1",
      "value": Object {
        "type": "number",
        "value": 9,
      },
    },
    Object {
      "adr": "C1",
      "value": Object {
        "formula": "sum(A1:B1)",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#REF!",
        },
      },
    },
    Object {
      "adr": "D1",
      "value": Object {
        "type": "number",
        "value": 13,
      },
    },
    Object {
      "adr": "A2",
      "value": Object {
        "formula": "A1+1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#REF!",
        },
      },
    },
    Object {
      "adr": "B2",
      "value": Object {
        "formula": "#REF!+10",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#REF!",
        },
      },
    },
    Object {
      "adr": "C2",
      "value": Object {
        "formula": "B1+6",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 15,
        },
      },
    },
  ],
  Array [],
]
`;

exports[`excel-formula-engine insertRows works 1`] = `
Array [
  Object {
    "address": "A5",
    "after": Object {
      "formula": "sum(A3:C4)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 7,
      },
    },
    "before": Object {
      "formula": "sum(a1:c2)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 7,
      },
    },
    "type": "update",
  },
  Object {
    "address": "C5",
    "after": Object {
      "formula": "sum(A3:B5)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 19,
      },
    },
    "before": Object {
      "formula": "sum(a1:b3)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 19,
      },
    },
    "type": "update",
  },
  Object {
    "address": "B3",
    "after": Object {
      "formula": "A3+1",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 2,
      },
    },
    "before": Object {
      "formula": "a1+1",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 2,
      },
    },
    "type": "update",
  },
  Object {
    "address": "C3",
    "after": Object {
      "formula": "B3+2",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 4,
      },
    },
    "before": Object {
      "formula": "b1+2",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 4,
      },
    },
    "type": "update",
  },
  Object {
    "address": "D3",
    "after": Object {
      "formula": "C3+3",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 7,
      },
    },
    "before": Object {
      "formula": "c1+3",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 7,
      },
    },
    "type": "update",
  },
  Object {
    "address": "E3",
    "after": Object {
      "formula": "C3+4",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 8,
      },
    },
    "before": Object {
      "formula": "c1+4",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 8,
      },
    },
    "type": "update",
  },
  Object {
    "address": "F3",
    "after": Object {
      "formula": "D3+H3",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#CYCLE!",
      },
    },
    "before": Object {
      "formula": "d1+h1",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#CYCLE!",
      },
    },
    "type": "update",
  },
  Object {
    "address": "G3",
    "after": Object {
      "formula": "F3",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#CYCLE!",
      },
    },
    "before": Object {
      "formula": "f1",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#CYCLE!",
      },
    },
    "type": "update",
  },
  Object {
    "address": "H3",
    "after": Object {
      "formula": "G3",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#CYCLE!",
      },
    },
    "before": Object {
      "formula": "g1",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#CYCLE!",
      },
    },
    "type": "update",
  },
  Object {
    "address": "I3",
    "after": Object {
      "formula": "F3",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#CYCLE!",
      },
    },
    "before": Object {
      "formula": "f1",
      "type": "formula",
      "value": Object {
        "message": "",
        "type": "error",
        "value": "#CYCLE!",
      },
    },
    "type": "update",
  },
  Object {
    "address": "A6",
    "after": Object {
      "formula": "A5+1",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 8,
      },
    },
    "before": Object {
      "formula": "a3+1",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 8,
      },
    },
    "type": "update",
  },
  Object {
    "address": "B6",
    "after": Object {
      "formula": "D3+10",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 17,
      },
    },
    "before": Object {
      "formula": "d1+10",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 17,
      },
    },
    "type": "update",
  },
  Object {
    "address": "C6",
    "after": Object {
      "formula": "B5+6",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 15,
      },
    },
    "before": Object {
      "formula": "b3+6",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 15,
      },
    },
    "type": "update",
  },
]
`;

exports[`excel-formula-engine insertRows works 2`] = `
Array [
  Array [
    Object {
      "adr": "A3",
      "value": Object {
        "type": "number",
        "value": 1,
      },
    },
    Object {
      "adr": "B3",
      "value": Object {
        "formula": "A3+1",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 2,
        },
      },
    },
    Object {
      "adr": "C3",
      "value": Object {
        "formula": "B3+2",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 4,
        },
      },
    },
    Object {
      "adr": "D3",
      "value": Object {
        "formula": "C3+3",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 7,
        },
      },
    },
    Object {
      "adr": "E3",
      "value": Object {
        "formula": "C3+4",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 8,
        },
      },
    },
    Object {
      "adr": "F3",
      "value": Object {
        "formula": "D3+H3",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "G3",
      "value": Object {
        "formula": "F3",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "H3",
      "value": Object {
        "formula": "G3",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "I3",
      "value": Object {
        "formula": "F3",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "A5",
      "value": Object {
        "formula": "sum(A3:C4)",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 7,
        },
      },
    },
    Object {
      "adr": "B5",
      "value": Object {
        "type": "number",
        "value": 9,
      },
    },
    Object {
      "adr": "C5",
      "value": Object {
        "formula": "sum(A3:B5)",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 19,
        },
      },
    },
    Object {
      "adr": "D5",
      "value": Object {
        "type": "number",
        "value": 13,
      },
    },
    Object {
      "adr": "A6",
      "value": Object {
        "formula": "A5+1",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 8,
        },
      },
    },
    Object {
      "adr": "B6",
      "value": Object {
        "formula": "D3+10",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 17,
        },
      },
    },
    Object {
      "adr": "C6",
      "value": Object {
        "formula": "B5+6",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 15,
        },
      },
    },
  ],
  Array [],
]
`;

exports[`excel-formula-engine recompute works 1`] = `
Array [
  Object {
    "address": "C1",
    "after": Object {
      "type": "number",
      "value": 10,
    },
    "before": Object {
      "formula": "b1+2",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 4,
      },
    },
    "type": "update",
  },
  Object {
    "address": "A3",
    "after": Object {
      "formula": "sum(a1:c2)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 13,
      },
    },
    "before": Object {
      "formula": "sum(a1:c2)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 7,
      },
    },
    "type": "update",
  },
  Object {
    "address": "A4",
    "after": Object {
      "formula": "a3+1",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 14,
      },
    },
    "before": Object {
      "formula": "a3+1",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 8,
      },
    },
    "type": "update",
  },
  Object {
    "address": "C3",
    "after": Object {
      "formula": "sum(a1:b3)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 25,
      },
    },
    "before": Object {
      "formula": "sum(a1:b3)",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 19,
      },
    },
    "type": "update",
  },
  Object {
    "address": "E1",
    "after": Object {
      "formula": "c1+4",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 14,
      },
    },
    "before": Object {
      "formula": "c1+4",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 8,
      },
    },
    "type": "update",
  },
  Object {
    "address": "D1",
    "after": Object {
      "formula": "c1+3",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 13,
      },
    },
    "before": Object {
      "formula": "c1+3",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 7,
      },
    },
    "type": "update",
  },
  Object {
    "address": "B4",
    "after": Object {
      "formula": "d1+10",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 23,
      },
    },
    "before": Object {
      "formula": "d1+10",
      "type": "formula",
      "value": Object {
        "type": "number",
        "value": 17,
      },
    },
    "type": "update",
  },
]
`;

exports[`excel-formula-engine recompute works 2`] = `
Array [
  Array [
    Object {
      "adr": "A1",
      "value": Object {
        "type": "number",
        "value": 1,
      },
    },
    Object {
      "adr": "B1",
      "value": Object {
        "formula": "a1+1",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 2,
        },
      },
    },
    Object {
      "adr": "C1",
      "value": Object {
        "type": "number",
        "value": 10,
      },
    },
    Object {
      "adr": "D1",
      "value": Object {
        "formula": "c1+3",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 13,
        },
      },
    },
    Object {
      "adr": "E1",
      "value": Object {
        "formula": "c1+4",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 14,
        },
      },
    },
    Object {
      "adr": "F1",
      "value": Object {
        "formula": "d1+h1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "G1",
      "value": Object {
        "formula": "f1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "H1",
      "value": Object {
        "formula": "g1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "I1",
      "value": Object {
        "formula": "f1",
        "type": "formula",
        "value": Object {
          "message": "",
          "type": "error",
          "value": "#CYCLE!",
        },
      },
    },
    Object {
      "adr": "A3",
      "value": Object {
        "formula": "sum(a1:c2)",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 13,
        },
      },
    },
    Object {
      "adr": "B3",
      "value": Object {
        "type": "number",
        "value": 9,
      },
    },
    Object {
      "adr": "C3",
      "value": Object {
        "formula": "sum(a1:b3)",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 25,
        },
      },
    },
    Object {
      "adr": "D3",
      "value": Object {
        "type": "number",
        "value": 13,
      },
    },
    Object {
      "adr": "A4",
      "value": Object {
        "formula": "a3+1",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 14,
        },
      },
    },
    Object {
      "adr": "B4",
      "value": Object {
        "formula": "d1+10",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 23,
        },
      },
    },
    Object {
      "adr": "C4",
      "value": Object {
        "formula": "b3+6",
        "type": "formula",
        "value": Object {
          "type": "number",
          "value": 15,
        },
      },
    },
  ],
  Array [],
]
`;
